File: keet-expo-modules/ios/KeetVideoUtilsModule.swift
import ExpoModulesCore

public class KeetVideoUtilsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("KeetVideoUtilsModule")

    AsyncFunction("getSize") { (videoUri: URL, promise: Promise) in
      guard let size = getVideoDimensions(from: videoUri) else {
        promise.resolve(nil)
        return
      }

      promise.resolve([
        "width": size.width,
        "height": size.height,
      ])
    }
  }

  func getVideoDimensions(from url: URL) -> CGSize? {
    let asset = AVURLAsset(url: url)
    let tracks = asset.tracks(withMediaType: .video)
    let track = tracks.first

    return track?.naturalSize
  }
}

File: keet-expo-modules/android/src/main/java/io/keet/KeetVideoUtilsModule.kt
package io.keet

import android.media.MediaMetadataRetriever
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.FileInputStream

class KeetVideoUtilsModule: Module() {
  override fun definition() = ModuleDefinition {
    Name("KeetVideoUtilsModule")

    AsyncFunction("getSize") { videoUri: String, promise: Promise ->
      println(videoUri)
      CoroutineScope(Dispatchers.IO).launch {
        promise.resolve(getVideoDimensions(videoUri))
      }
    }
  }

  private suspend fun getVideoDimensions(videoUrl: String): Map<String, Int>? {
    return withContext(Dispatchers.IO) {
      try {
        val retriever = MediaMetadataRetriever()

        if (videoUrl.startsWith("file://") || videoUrl.startsWith("/")) {
          val inputStream = FileInputStream(videoUrl.replace("file://", ""));
          retriever.setDataSource(inputStream.fd)
        } else {
          retriever.setDataSource(videoUrl)
        }

        val width = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH)?.toInt()
        val height = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT)?.toInt()
        retriever.release()

        if (width != null && height != null) {
          mapOf(
            "width" to width,
            "height" to height,
          )
        } else {
          null
        }
      } catch (e: Exception) {
        e.printStackTrace()
        null
      }
    }
  }
}
